### DISCUSSION TITLE
Allow sending of arbitrary data via `unauthorized`.

### DISCUSSION THREAD

**Discussion by simoncave** [vercel/next.js#73775] [Ideas]
### Goals

Allow context to be sent from `unauthorized()` to `unauthorized.js`.

### Background

Next.js 15.1 introduces the [`unauthorized`](https://nextjs.org/docs/canary/app/api-reference/functions/unauthorized) and [`forbidden`](https://nextjs.org/docs/canary/app/api-reference/functions/forbidden) functions, available behind the [`authInterrupts`](https://nextjs.org/docs/canary/app/api-reference/config/next-config-js/authInterrupts) configuration option.

The documentation for `unauthorized` includes an example titled ["Displaying login UI to unauthenticated users"](https://nextjs.org/docs/canary/app/api-reference/functions/unauthorized#displaying-login-ui-to-unauthenticated-users). In this example, a server component checks to see if the user is authenticated and, if not, calls `unauthorized()`. This interrupts rendering of the component by throwing an error, which is captured in [`unauthorized.js`](https://nextjs.org/docs/canary/app/api-reference/file-conventions/unauthorized), which renders log-in UI as a fallback.

This works for simple authorization schemes, where you're either authorized (`200`) or not (`401`). But in some authorization schemes, authorization comes with variable levels of assurance based on _how_ you authenticated. This means that there can be more than two outcomes:

1. `200 OK` — You're authorized.
2. `401 Unauthorized` — You're not authorized, so you need to log-in.
3. `401 Unauthorized` — You're not authorized _enough_, so you need to _step-up_ by using a stronger form of authentication (e.g. a passkey on a physical security key) [^1].

[^1]: It might seem like scenario 3 should be a 403, but the devil is in the details. A 403 error is about a subject's relationship to a resource. It means _"you're not allowed to do that"_. Scenario 3 means _"you are allowed to do that, but before I let you do it, I need more assurance that you are who you say you are"_.

If Next.js is in full control of the authorization scheme, then it should be possible to distinguish between scenarios 2 and 3 from `unauthorized.js`. But if the authorization violation occurred in an upstream service which is returning a 401 response to Next.js, then we need to be able to provide an argument to `unauthorized` that we can read from `unauthorized.js`.

### Proposal

The `unauthorized` function should accept a single argument which is either:

  - Ideally: Any arbitrary value which can be serialized by React.
  - At a minimum: Any string.

###### server-actions.ts

```tsx
"use server";

import { unauthorized } from "next/navigation";

export async function sendPayment(options: { recipient: string, amount: number }) {
  const authnToken = (await cookies()).get("AuthnToken");

  const response = await fetch("https://payments-service.example", {
    body: JSON.stringify(
    headers: {
      authorization: `Bearer ${authnToken)`,
    },
  });

  if (response.status === 401) {
    const stepUpToken = response.headers.get("x-authz-step-up-token");
    unauthorized(stepUpToken);
  }

  if (!response.ok) {
    throw new Error("Payment failed.");
  }
  
  return { ok: true };
}
```

The `unauthorized.js` file should receive the argument as a prop.

###### unauthorized.ts

```tsx
import Login from "@/app/components/Login";
import StepUp from "@/app/components/StepUp";
 
export default function UnauthorizedPage(props: { detail: unknown }) {
 const stepUpToken = props.detail;

  return (
    <main>
      <h1>401 - Unauthorized</h1>
      {stepUpToken ? <StepUp token={stepUpToken} /> : <Login />}
    </main>
  )
}
```

> [!NOTE]
> This feature request focusses on `unauthorized`, but what makes sense for `unauthorized` may also make sense for `forbidden` and even `notFound`.

**Comment by colinclerk** [comment]
Clerk would also like the ability to pass reasons through error responses, with the same immediate goal of improving DX for "Step Up" auth (which we call "reverification" of signed in users).

We would also like the ability to handle server action errors from the client, instead of just from unauthorized.ts, but falling back to unauthorized.ts is a reasonable start.

**Reply by simoncave** [reply]
Any thoughts on the API? Would you change anything about what I proposed?

**Comment by feedthejim** [comment]
Hey @simoncave, thanks for writing this up. Could you move close this and move this over to the RFC discussion here https://github.com/vercel/next.js/discussions/73753 ? I will coalesce all feedback there and bring it back to the team.

**Reply by simoncave** [reply]
👍🏻 → https://github.com/vercel/next.js/discussions/73753#discussioncomment-11553741

**Comment by simoncave** [comment]
See https://github.com/vercel/next.js/discussions/73753#discussioncomment-11553741
